File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.py
import sys
from collections import deque
def bfs(n,m,grid):
    directions=[(-1,0),(1,0),(0,-1),(0,1)]
    visited=[[[False]*2 for _ in range(m)] for _ in range(n)]
    que=deque([(0,0,0,1)])
    visited[0][0][0]=True
    while que:
        y,x,wall_broken,dist=que.popleft()
        if y==n-1 and x==m-1:
            return dist
        for dy,dx in directions:
            # ny=y+dy
            # nx=x+dx
            ny,nx=y+dy,x+dx
            if 0<=ny<n and 0<=nx<m:
                if grid[ny][nx]==0 and not visited[ny][nx][wall_broken]:
                    visited[ny][nx][wall_broken]=True
                    que+=[(ny,nx,wall_broken,dist+1)]
                elif grid[ny][nx]==1 and wall_broken==0 and not visited[ny][nx][1]:
                    visited[ny][nx][1]=True
                    que+=[(ny,nx,1,dist+1)]
    return -1
input=lambda:sys.stdin.readline().rstrip()
n,m=map(int,input().split())
grid=[[int(a) for a in input()] for _ in range(n)]
print(bfs(n,m,grid))

File: 백준/Gold/2206. 벽 부수고 이동하기/벽 부수고 이동하기.cc
#include <deque>
#include <iostream>
#include <vector>
using namespace std;
int bfs(int n, int m, vector<vector<int>> &grid) {
    int y, x, ny, nx, wall_broken, dist;
    vector<vector<int>> directions = {{-1, 0}, {1, 0}, {0, -1}, {0, 1}};
    vector<vector<vector<bool>>> visited(n, vector<vector<bool>>(m, vector<bool>(2, false)));
    deque<vector<int>> que;
    que.push_back({0, 0, 0, 1});
    visited[0][0][0] = true;
    while (!que.empty()) {
        y = que.front()[0], x = que.front()[1], wall_broken = que.front()[2], dist = que.front()[3];
        que.pop_front();
        if (y == n - 1 && x == m - 1) {
            return dist;
        }
        for (int i = 0; i < directions.size(); i++) {
            ny = y + directions[i][0];
            nx = x + directions[i][1];
            if (0 <= ny && ny < n && 0 <= nx && nx < m) {
                if (grid[ny][nx] == 0 && !visited[ny][nx][wall_broken]) {
                    visited[ny][nx][wall_broken] = true;
                    que.push_back({ny, nx, wall_broken, dist + 1});
                } else if (grid[ny][nx] == 1 && wall_broken == 0 && !visited[ny][nx][1]) {
                    visited[ny][nx][1] = true;
                    que.push_back({ny, nx, 1, dist + 1});
                }
            }
        }
    }
    return -1;
}
int main(void) {
    int n, m;
    cin >> n >> m;
    vector<string> t(n);
    vector<vector<int>> grid(n, vector<int>(m, 0));
    for (int i = 0; i < n; i++)
        cin >> t[i];
    for (int i = 0; i < n; i++)
        for (int j = 0; j < m; j++)
            grid[i][j] = t[i][j] - 48;
    cout << bfs(n, m, grid);
    return 0;
}